### ISSUE TITLE
Markdown support?

### ISSUE THREAD

**Issue by joeaudette** [joeaudette/cloudscribe.SimpleContent#199]
A few people have asked me in person about markdown support for SimpleContent so I wanted to create an issue to discuss it.

In general I think it is mainly developers who like to write in markdown not typical non-technical site authors. So there is perhaps a limited audience who this would be appealing for, but that doesn't mean it isn't worth doing if there is sufficient interest.

### What should the implementation be like?

From a quick look around the web it seems that most of the current blogs/content engines that use markdown are mainly static site generators so the html is generated ahead of time not rendered from markdown at runtime, but the source of truth is the markdown files. SimpleContent is not a static site generator so we would indeed be converting from markdown to html at runtime which I don't see as a problem. There are plenty of static site generators out there already and that is not what SimpleContent is meant to be.

If we were to implement use of markdown as an option, then the next conversation is what that should look like. Is it enough to be able to edit using markdown syntax in the browser or is it important that the blog posts for example are stored as markdown files on disk or would it be ok to continue using xml or json and just storing the post body within those formats as markdown?

It would be theoretically possible to serialize posts as markdown with a custom post serializer for NoDb in order to persist the files as markdown. There may be some challenges for how meta data about the post would be stored within the markup but not necessarily rendered when the markup is converted to html for display. One big disadvantage of that idea (serializing posts to markdown) is that it would mean that markdown could ONLY be used if using NoDb storage, it could not be used with a database back end unless we just store the post body as markdown in the database. Storing files as markdown on disk would have the advantage that other markdown editors or desktop tools could be used to create content offline. Another advantage of storing the posts on disk as markdown is that would improve the possibilities for people to migrate to SimpleContent from static site generators just by dropping in their markdown files, but it may also be subject to variations in markdown extensions for such things as metadata or other non-standard extensions for markdown used in specific static site generators. So there are definite pros and cons to be considered for both possible approaches.

One of the other challenges that I can think of is is using markdown to edit posts in the browser then we still need a way to upload images and then insert the markup for the image into the markdown editor. So we would need some kind of javascript to facilitate the editing experience. Probably could support dropping images into the editor in a fairly straightforward way. Having a toolbar to browse existing images on the server would be nice. Maybe there is an open source javascript editor for markdown that we could use like the one here on github. We should explore what is available and list them on this thread. From a quick search this one looks great https://github.com/sparksuite/simplemde-markdown-editor and already has events such as drop that we could handle for image upload.

There are also other challenges to consider such as how to implement code syntax highlighting which we already have as a plugin for CKeditor. How to implement easy embedding of youtube videos and other sources such as we already have in CKeditor.

**Comment by FinnurHrafn** [comment]
I don't see a significant need for markdown support in what I plan do to with cloudscribe. But then again my needs are modest. A few simple low-traffic sites, partly edited by me or some people with no knowledge of markdown. I also see opportunities for cloudscribe in 2-3 custom web apps that are in preparation, but again no need for markdown. 
I value the simplicity of cloudscribe, this is a significant addition to the complexity and enlarges the test matrix for cloudscribe considerably.

**Comment by juergen-de** [comment]
For me, SimpleContent is a limited, lightweight CMS Module for Cloudscribe. But with switching to source view, I can use html and apply bootstrap classes which is a game changer for me. Switching to a markdown editor would probably be a step backwards and as Finnur said, add unnecessary complexity.
I see points in SimpleContent where you could get far more bang for the buck.

**Comment by CrispinF** [comment]
For me, no markdown needed or justified. I agree with your comment Joe that it is mostly devs who like it (and I've no idea why!).

**Comment by joeaudette** [comment]
Thanks all for the feedback so far. I will only consider it if there is demand for it, and since I "think" only developers would be the ones interested in it I think having this github issue will be a good way to measure interest since I assume it is mostly developers reading github issues. I will keep this issue open for an indefinite period to see if over time more interest is expressed.

I do want to wash away any fear that implementing support for markdown would introduce a lot of complexity as I do not think it would. I have been thinking more about it since first posting the issue and I kind of have a vision for how it could be implemented which I will elaborate on.

## Envisioned Implementation Outline.

In SimpleContent ProjectSettings a new property would be introduced for "DefaultContentType" which would of course have a default value "html" and could be surfaced in the UI as a dropdown with the choices html or markdown. This property would only affect how new pages and posts would be created.

In Post and Page classes (IPost and IPage) a ContentType property would be added which again would default as "html" but when new content is created it would be set according the project setting.

Then in the display and edit views it would simply be a matter of using different partial views depending on the content type, so existing html content could still be used and edited even if someone later changed the default to markdown.

That is really all the changes needed so either html or markdown could be stored in the existing storage for the body content of posts and pages and it would even work with EF database storage.

If there were demand to actually store the files as markdown.md files that could be in a separate class library named cloudscribe.SimpleContent.NoDb.Markdown
In that library is where a custom implementation of IStringSerializer of Post  and IStringSerializer of Page could be implemented. Currently we have an [XmlSerializer for posts](https://github.com/joeaudette/cloudscribe.SimpleContent/blob/master/src/cloudscribe.SimpleContent.Storage.NoDb/PostXmlSerializer.cs) and we just use the [default json serializer from NoDb](https://github.com/joeaudette/NoDb/blob/master/src/NoDb/StringSerializer.cs) for Page

But different serializers that serialize back and forth from markdown could be implemented and injected from startup to override the default serializers and they could be in a separate optional nuget

So it all adds up to 3 new model properties, a few new optional partial views and 2 new optional interface implementations for serialization.

Actually I'm hoping that if some developers do express interest in this that they would also be interested in implementing the serialization interfaces to support it.

**Comment by joeaudette** [comment]
@DannyAllegrezza wanted to bring this to your attention, if I recall you were one of the people who asked me about markdown support.

**Comment by DannyAllegrezza** [comment]
Hey @joeaudette,

Thanks for bringing this to my attention. I have also been thinking about this topic and after playing around with the project some I am of a different opinion now. Like you mentioned:

> There are plenty of static site generators out there already and that is not what SimpleContent is meant to be.

That said, I have found the WYSIWYG to be pretty powerful. Markdown might be a "nice to have" addition, but certainly not required by any means.

**Comment by moljac** [comment]
@joeaudette 

Right question for me, because I started to dig through the code to find and maybe add MD support (to contribute). In last few days I searched through the code to see which WYSIWYG HTML editor was used.

My background: I used Composite C1 mainly because of No-Database concept which is ideal for small business (like mine used to be and for most of the people I deal/dealt with).  Now I want to use ASP.net Core and seems like Composite C1 will not be ready for quite some time. I did contribute to C1 few things. I have to recall what I did and my web is dead right now to check it out. I'm moving everything to Azure.

I work on Xamarin Components Team (MSFT FTE) and web (ASP.net) is not my home/expertise, but I want to get deeper into that. On Components team (and Xamarin docs) everything is MD, even Workbooks. I keep my notes in MD. Notes about everything (Xamarin.Forms, Azure, Xamarin Bindings) and I need them often... And sometimes I need to share note with some devs. Right now I point to github public repo and MD doc, but for SEO I would like to have it on my site/blog. Being healthy-lazy I don't want to generate HTML manually from my MD. I will do it in the initial phase, but later I want automagical process. That is the reason I pinged you on gitter about page events.

So, my milestones (I need those to publish my old and new content to the new website and to increase SEO visibility for it):

1.  add *automagic* SEO ping feature for sitemaps (pinging google, bing, etc when page is added, deleted, updated).
2. add FileSystemWatcher to watch MD folders and generate static HTML folders (even better cloudscribe NoDB dataformat), so I can add MD content via copy paste into folder, or WebAPI (mobile client planned too - we started one for C1, but never had time to finish it up). 
3. add support for HTML/Razor/whatever templating support (template with js/css where MD generated HTML will be injected)

These are my use-cases and I think from a lot of the people on Xamarin team (Components, Docs, etc...). With this we would be able to represent docs, Workbooks everything...

### Comments based on really fast reading

> In general I think it is mainly developers who like to write in markdown not typical non-technical site authors. So there is perhaps a limited audience who this would be appealing for, but that doesn't mean it isn't worth doing if there is sufficient interest.

Imagine mobile client pushing some news item (post) to CMS (cloudscribe). I will not write HTML on iPhone or Android device (sorry, not worth it). MD would be ideal, even for  "typical non-technical" people. Add after than social media generator that publishes known url (because it landed in CMS 1st) to FB, twitter, g+. SEO on stereoids.

> property would be introduced for "DefaultContentType" which would of course have a default value "html" and could be surfaced in the UI as a dropdown with the choices html or markdown. This property would only affect how new pages and posts would be created.

This is what I had in mind to enhance "Add/Edit Page" admin page.

> If there were demand to actually store the files as markdown.md files that could be in a separate class library named cloudscribe.SimpleContent.NoDb.Markdown

I think there would be demand to have files saved as MD. FileSystemWatcher would be nice addition. Just an idea: either content can be Url to the content (slow because of test), or some property that defines that conent is located on FS.

> In that library is where a custom implementation of IStringSerializer of Post and IStringSerializer of Page could be implemented. Currently we have an XmlSerializer for posts and we just use the default json serializer from NoDb for Page

It would be good to investigate additional serialization formats and frameworks, just for the future use (middleware/WebAPI) like Bond, Thrift, Avro, Protobuffers, ... I know this does not bring too much for the MD/HTML content, but for the traffic etc...

> But different serializers that serialize back and forth from markdown could be implemented and injected from startup to override the default serializers and they could be in a separate optional nuget

This is it. By default JSON/XML and other as optional nugets.

> drag&drop  (images/content)

Overkill maybe (I will write MD in the txt files and save them somewhere, separate repo, folder, ...)

I have few PoC projects I use internally where I need backend and cloudscribe looks promising.

**Comment by joeaudette** [comment]
@moljac interesting comments. 

You use case of writing news on mobile client. I wonder how many people would really write content on a small device like a phone, a tablet seems more likely. But users don't have to write raw html anyway, the wysiwyg should work on mobile, but also seems like one could implement a mobile app that uses metaweblog api similar as Open Live Writer and could write content with that. Or a mobile app could be implemented that writes content in markdown but then converts it to html before pushing it to the server. There was once an iphone/ipad app named blogsy that I used before with metaweblog api, but it was not that friendly to use and I think the developer stopped developing it.

There are lots of tools for both server side and client side converting MD to HTML. A project I worked on recently I used [showdown](https://github.com/showdownjs/showdown) and unobtrusive javascript to load some MD files from github and render them as html fragments in an existing page. It was trivial to do. I add a div with a data- attribute with the url to the github md file like this:

    <div data-mdurl="someurltoamdfile"></div>

then just include the showdown js and a js file with unobtrusive js like this:

    $(function () {
        var $elems = $('div[data-mdurl]');
        if ($elems) {
            var converter = new showdown.Converter();
            $elems.each(function (index, ele) {
                var $url = $(ele).data('mdurl');
                $.get($url, function (data) {
                   ele.innerHTML = converter.makeHtml(data);
                }, 'text');
            });
         }
    });

it will detect the data attribute and render the md as html into the div with the attribute.

Using that technique one could easily render remote md content into simplecontent pages or posts with no changes needed in simplecontent.

supporting MD as a serialization format and storing the files as md would require md with some meta data extensions to handle other non-content properties of the model. I think it would be significant work and though you seem interested in it I'm still thinking you fall into the small niche of developers who would be interested in it.

**Comment by joeaudette** [comment]
I just added a document because it occurred to me that other might find this solution useful https://www.cloudscribe.com/docs/using-markdown

**Comment by joeaudette** [comment]
@moljac to be clear, if you are interested in doing the work to implement a markdown editor and/or serialization to and from markdown. I am certainly willing to add the needed content type field on post and page and defaultcontenttype field on project settings. I think just those things would be needed directly in the project, and your work could be in separate projects/nugets

**Comment by joeaudette** [comment]
@all I've implemented support for editing the content body of posts and pages as markdown. I even implemented support for dropping images into the markdown editor with automatic upload and resizing just like in the html editor.

So the only thing left to do is implement an option with NoDb storage to store the files on disk as markdown, which I believe could be done using [markdig](https://github.com/lunet-io/markdig). It supports YAML frontmatter which I think can be used for storing the non-content model properties as meta data within the markdown. markdig is also used currently for rendering the markdown as html.

I've updated to documentaiton to show how to enable markdown content type.

https://www.cloudscribe.com/docs/using-markdown
